### PR TITLE
SDI-318 handle async clients

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Handle async clients on the non-reactive API and add async client integration tests for both APIs.

### CODE REVIEW

File: hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt [hunk 1]
```diff
@@ -12,13 +13,15 @@ import org.springframework.web.server.ResponseStatusException
 
 @RestController
 @RequestMapping("/queue-admin")
-class HmppsQueueResource(private val hmppsQueueService: HmppsQueueService) {
+class HmppsQueueResource(private val hmppsQueueService: HmppsQueueService, private val hmppsAsyncQueueService: HmppsAsyncQueueService) {
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueResource.kt:16]: The main goal of this PR is to bring the non-reactive API up to date with the reactive API by handling asynchronous SQS clients.

File: test-app-async/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagenameasync/integration/HmppsQueueResourceTest_asyncClient.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import org.junit.jupiter.api.Test
+import org.mockito.kotlin.verify
+import org.springframework.http.MediaType
+import software.amazon.awssdk.services.sqs.model.SendMessageRequest
+import uk.gov.justice.digital.hmpps.hmppstemplatepackagenameasync.service.EventType
+import uk.gov.justice.digital.hmpps.hmppstemplatepackagenameasync.service.HmppsEvent
+import uk.gov.justice.digital.hmpps.hmppstemplatepackagenameasync.service.Message
+import uk.gov.justice.digital.hmpps.hmppstemplatepackagenameasync.service.MessageAttributes
+
+class HmppsQueueResourceTest_asyncClient : IntegrationTestBase() {
```
Comment (mikehalmamoj) [inline test-app-async/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagenameasync/integration/HmppsQueueResourceTest_asyncClient.kt:17]: Added some integration tests for async SQS clients to both the non-reactive and reactive test apps.

File: test-app-async/src/test/resources/application-test.yml [hunk 3]
```diff
@@ -32,6 +32,11 @@ hmpps.sqs:
       queueName: ${random.uuid}
       subscribeTopicId: outboundtopic
       subscribeFilter: '{"eventType":[ "offender.movement.reception"] }'
+    asyncqueue:
+      dlqName: ${random.uuid}
+      queueName: ${random.uuid}
+      asyncDlqClient: true
+      asyncQueueClient: true
```
Comment (mikehalmamoj) [inline test-app-async/src/test/resources/application-test.yml:39]: Added a new queue with asynchronous SQS clients to test against.

File: hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsSqsConfiguration.kt [hunk 4]
```diff
@@ -15,7 +15,7 @@ import org.springframework.jms.annotation.EnableJms
 @Configuration
 @EnableConfigurationProperties(HmppsSqsProperties::class)
 @EnableJms
-@AutoConfigureBefore(HealthEndpointAutoConfiguration::class)
+@AutoConfigureBefore(WebFluxAutoConfiguration::class)
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsSqsConfiguration.kt:18]: The async queue was being wired up after the Health Registry - something to do with the dependency graph between beans. So we need to jump in and create our beans as early as possible...

File: test-app-async/src/main/resources/application-localstack.yml [hunk 5]
```diff
@@ -11,6 +11,11 @@ hmpps.sqs:
       dlqName: async-outbound-dlq
       subscribeTopicId: outboundtopic
       subscribeFilter: '{"eventType":[ "offender.movement.reception"] }'
+    asyncqueue:
+      dlqName: async-async-dlq
+      queueName: async-async-queue
+      asyncDlqClient: true
+      asyncQueueClient: true
```
Comment (mikehalmamoj) [inline test-app-async/src/main/resources/application-localstack.yml:18]: Adds a queue with an async client. Note async message listeners not currently supported so just used for testing the queue resource endpoints (purge queue etc.) The strange name is to keep with existing naming convention - will probably change to `reactive-async-dlq` when I do some refactoring.

Comment (mikehalmamoj) [inline test-app-async/src/main/resources/application-localstack.yml:18]: The async client for running the test app locally. The strange name is to keep with existing naming convention - will probably change to `reactive-async-dlq` when I do some refactoring.

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**